File: packages/shared/src/utils/thinking.ts
import type { Dispatch, SetStateAction } from "react";
import { AIMessage, type BaseMessage } from "@langchain/core/messages";
import { THINKING_MODELS } from "../models.js";

type ThinkingAndResponseTokens = {
  thinking: string;
  response: string;
};

/**
 * Extracts thinking and response content from a text string containing XML-style think tags.
 * Designed to handle streaming text where tags might be incomplete.
 *
 * @param text - The input text that may contain <think> tags
 * @returns An object containing:
 *   - thinking: Content between <think> tags, or all content after <think> if no closing tag
 *   - response: All text outside of think tags
 *
 * @example
 * // Complete tags
 * extractThinkingAndResponseTokens('Hello <think>processing...</think>world')
 * // Returns: { thinking: 'processing...', response: 'Hello world' }
 *
 * // Streaming/incomplete tags
 * extractThinkingAndResponseTokens('Hello <think>processing...')
 * // Returns: { thinking: 'processing...', response: 'Hello ' }
 *
 * // No tags
 * extractThinkingAndResponseTokens('Hello world')
 * // Returns: { thinking: '', response: 'Hello world' }
 */
export function extractThinkingAndResponseTokens(
  text: string
): ThinkingAndResponseTokens {
  const thinkStartTag = "<think>";
  const thinkEndTag = "</think>";

  const startIndex = text.indexOf(thinkStartTag);

  // No thinking tag found
  if (startIndex === -1) {
    return {
      thinking: "",
      response: text.trim(),
    };
  }

  const afterStartTag = text.substring(startIndex + thinkStartTag.length);
  const endIndex = afterStartTag.indexOf(thinkEndTag);

  // If no closing tag, all remaining text is thinking
  if (endIndex === -1) {
    return {
      thinking: afterStartTag.trim(),
      response: text.substring(0, startIndex).trim(),
    };
  }

  // We have both opening and closing tags
  const thinking = afterStartTag.substring(0, endIndex).trim();
  const response = (
    text.substring(0, startIndex) +
    afterStartTag.substring(endIndex + thinkEndTag.length)
  ).trim();

  return {
    thinking,
    response,
  };
}

type HandleRewriteArtifactThinkingModelParams = {
  newArtifactContent: string;
  setMessages: Dispatch<SetStateAction<BaseMessage[]>>;
  thinkingMessageId: string;
};

/**
 * Handles the rewriting of artifact content by processing thinking tokens and updating messages state.
 * This function extracts thinking and response tokens from the new artifact content, updates the message
 * state with thinking tokens if present, and returns the response content.
 *
 * @param {Object} params - The parameters for handling artifact rewriting
 * @param {string} params.newArtifactContent - The new content to process for the artifact
 * @param {Dispatch<SetStateAction<BaseMessage[]>>} params.setMessages - State setter function for updating messages
 * @param {string} params.thinkingMessageId - Unique identifier for the thinking message to update or create
 * @returns {string} The extracted response content from the artifact
 */
export function handleRewriteArtifactThinkingModel({
  newArtifactContent,
  setMessages,
  thinkingMessageId,
}: HandleRewriteArtifactThinkingModelParams): string {
  const { thinking, response } =
    extractThinkingAndResponseTokens(newArtifactContent);

  if (thinking.length > 0) {
    setMessages((prevMessages) => {
      if (!thinkingMessageId) {
        console.error("Thinking message not found");
        return prevMessages;
      }

      const prevHasThinkingMsg = prevMessages.some(
        (msg) => msg.id === thinkingMessageId
      );

      const thinkingMessage = new AIMessage({
        id: thinkingMessageId,
        content: thinking,
      });

      if (prevHasThinkingMsg) {
        // The message exists, so replace it
        const newMsgs = prevMessages.map((msg) => {
          if (msg.id !== thinkingMessageId) {
            return msg;
          }
          return thinkingMessage;
        });
        return newMsgs;
      }
      // The message does not yet exist, so create it:
      return [...prevMessages, thinkingMessage];
    });
  }

  return response;
}

export function isThinkingModel(model: string): boolean {
  return THINKING_MODELS.some((m) => m === model);
}

File: agents-py/shared/src/utils/thinking.py
from typing import Dict, List, Optional, Tuple
from pydantic import BaseModel
from langchain_core.messages import AIMessage, BaseMessage
from shared.src.models import THINKING_MODELS

class ThinkingAndResponseTokens(BaseModel):
    thinking: str
    response: str

def extract_thinking_and_response_tokens(text: str) -> ThinkingAndResponseTokens:
    """
    Extracts thinking and response content from text containing XML-style think tags.
    Handles incomplete tags during streaming.

    Args:
        text: Input text potentially containing <think> tags

    Returns:
        ThinkingAndResponseTokens with separated content

    Example:
        >>> extract_thinking_and_response_tokens('Hello <think>processing...</think>world')
        ThinkingAndResponseTokens(thinking='processing...', response='Hello world')
    """
    think_start_tag = "<think>"
    think_end_tag = "</think>"

    start_idx = text.find(think_start_tag)
    if start_idx == -1:
        return ThinkingAndResponseTokens(thinking="", response=text.strip())

    after_start = text[start_idx + len(think_start_tag):]
    end_idx = after_start.find(think_end_tag)

    if end_idx == -1:
        return ThinkingAndResponseTokens(
            thinking=after_start.strip(),
            response=text[:start_idx].strip()
        )

    thinking = after_start[:end_idx].strip()
    response = (text[:start_idx] + after_start[end_idx + len(think_end_tag):]).strip()
    return ThinkingAndResponseTokens(thinking=thinking, response=response)

class HandleRewriteParams(BaseModel):
    new_content: str
    messages: List[BaseMessage]
    thinking_id: str

def handle_rewrite_artifact_thinking(params: HandleRewriteParams) -> Tuple[str, List[BaseMessage]]:
    """
    Processes thinking tokens in artifact content and updates messages.

    Returns:
        Tuple of (cleaned response content, updated messages list)
    """
    extracted = extract_thinking_and_response_tokens(params.new_content)
    new_messages = params.messages.copy()

    if extracted.thinking:
        # Update or create thinking message
        updated = False
        for i, msg in enumerate(new_messages):
            if msg.id == params.thinking_id:
                new_messages[i] = AIMessage(
                    id=params.thinking_id,
                    content=extracted.thinking
                )
                updated = True
                break
        
        if not updated:
            new_messages.append(AIMessage(
                id=params.thinking_id,
                content=extracted.thinking
            ))

    return extracted.response, new_messages

def is_thinking_model(model_name: str) -> bool:
    """Check if model is configured for thinking output"""
    return model_name in THINKING_MODELS 